backend/src/email/routes/tests/email-transactional.routes.test.ts

```diff
@@ -55,85 +92,107 @@ describe('POST /transactional/email/send', () => {
-      .post('/transactional/email/send')
+      .post(endpoint)
       .set('Authorization', `Bearer ${apiKey}`)
       .send({})
 
     expect(res.status).toBe(400)
   })
 
   test('Should send a message successfully', async () => {
-    const mockSendEmail = jest
-      .spyOn(EmailService, 'sendEmail')
-      .mockResolvedValue('message_id')
-
     const res = await request(app)
-      .post('/transactional/email/send')
+      .post(endpoint)
       .set('Authorization', `Bearer ${apiKey}`)
-      .send({
-        recipient: 'recipient@agency.gov.sg',
-        subject: 'subject',
-        body: '<p>body</p>',
-        from: 'Postman <donotreply@mail.postman.gov.sg>',
-        reply_to: 'user@agency.gov.sg',
-      })
+      .send(validApiCall)
 
     expect(res.status).toBe(202)
     expect(mockSendEmail).toBeCalledTimes(1)
+    const transactionalEmail = await EmailMessageTransactional.findOne({
+      where: { userId: userId.toString() },
+    })
+    expect(transactionalEmail).not.toBeNull()
+    expect(transactionalEmail).toMatchObject({
+      recipient: validApiCall.recipient,
+      from: validApiCall.from,
+      status: TransactionalEmailMessageStatus.Accepted,
+      errorCode: null,
+    })
+    expect(transactionalEmail?.params).toMatchObject({
+      subject: validApiCall.subject,
+      body: validApiCall.body,
+      from: validApiCall.from,
+      replyTo: validApiCall.reply_to,
+    })
   })
 
   test('Should send a message with valid custom from name', async () => {
-    const mockSendEmail = jest
-      .spyOn(EmailService, 'sendEmail')
-      .mockResolvedValue('message_id')
-
     const res = await request(app)
-      .post('/transactional/email/send')
+      .post(endpoint)
       .set('Authorization', `Bearer ${apiKey}`)
       .send({
-        recipient: 'recipient@agency.gov.sg',
-        subject: 'subject',
-        body: '<p>body</p>',
+        ...validApiCall,
         from: 'Hello <donotreply@mail.postman.gov.sg>',
         reply_to: user.email,
       })
 
     expect(res.status).toBe(202)
     expect(mockSendEmail).toBeCalledTimes(1)
+    const transactionalEmail = await EmailMessageTransactional.findOne({
+      where: { userId: userId.toString() },
+    })
+    expect(transactionalEmail).not.toBeNull()
+    expect(transactionalEmail).toMatchObject({
+      recipient: validApiCall.recipient,
+      from: 'Hello <donotreply@mail.postman.gov.sg>',
+      status: TransactionalEmailMessageStatus.Accepted,
+      errorCode: null,
+    })
+    expect(transactionalEmail?.params).toMatchObject({
+      subject: validApiCall.subject,
+      body: validApiCall.body,
+      from: 'Hello <donotreply@mail.postman.gov.sg>',
+      replyTo: user.email,
+    })
   })
 
+  // something I don't understand — is the verification of the from address mocked? or added to `email_from_address`?
+  // afaict, user.email is not added to `email_from_address` in the test setup, yet somehow the test passes
```

Comment:
to remove before merging in (genuinely not sure how it's done and whether there's a lesson here for how I should write my tests differently if so)

Reply:
Because we're mocking the email sending, ofc it passes

Reply:
Ohh I was confused why this tests passes and why the next test fails.

```TS
test('Should thrown an error with invalid custom from address', async () => {
    const res = await request(app)
      .post(endpoint)
      .set('Authorization', `Bearer ${apiKey}`)
      .send({
        ...validApiCall,
        from: 'Hello <invalid@agency.gov.sg>',
        reply_to: user.email,
      })

    expect(res.status).toBe(400)
    expect(mockSendEmail).not.toBeCalled()
})
```

But looking back I think I figured out what's going on.

I was expecting a db call to `EmailFromAddress` to check whether the the custom from address is in our database. This is used e.g. in `EmailMiddleware.existsFromAddress`, which is called for email campaigns.

In `email-transactional.routes.ts`, we call `emailMiddleware.isFromAddressAccepted`, which simply checks whether `(fromAddress !== userEmail && fromAddress !== defaultFromAddress)`. 

Per our previous discussion, we said it's enough to have an API for users to check that their custom from email address is acceptable and just throw them a 500 error if they pass the check above but fail at the AWS SES (see this [issue](https://github.com/opengovsg/postmangovsg/issues/1837)).

Reply:
Do you think this is something we should add? (Could be a separate PR, would handle the issue Kee Wei flagged as well)

Reply:
Yup we will need to handle that error, but not in this PR. Don't think it's a big issue (yet) since we're still manually onboarding our B2B users and not leaving then to self-service yet